### PR TITLE
Insights Manager refactoring

### PR DESCRIPTION
**Pull request by RagnarR**
Refactored Insights Manager and Network Adapter classes with Executor

### CODE REVIEW

File: mediation/src/main/java/net/pubnative/mediation/insights/PubnativeInsightsManager.java [hunk 1]
```diff
@@ -74,232 +82,51 @@ public synchronized static void trackData(Context context, String baseURL, Map<S
...
-            List<PubnativeInsightRequestModel> pendingList = getTrackingList(context, listKey);
-            if (pendingList == null) {
-                pendingList = new ArrayList<PubnativeInsightRequestModel>();
-            }
-            pendingList.add(model);
-            setTrackingList(context, listKey, pendingList);
+            sModel = new PubnativeInsightRequestModel(uriBuilder.build().toString(), dataModel);
+            sTrackingDataString = new Gson().toJson(sModel.dataModel);
+            sUrl = sModel.url;
+            sendTrackingDataToServer();
```
Comment (cerberillo) [inline mediation/src/main/java/net/pubnative/mediation/insights/PubnativeInsightsManager.java]: What if we launch several insights at the same time, won't sModel will be overwritten by following requests?

Comment (RagnarR) [inline mediation/src/main/java/net/pubnative/mediation/insights/PubnativeInsightsManager.java]: Good point. I think I need fixed it. 👍

**Comment by cerberillo** [comment]
It would be better for review if this PR contains only the planned changes.

To do this, I suggest to do a clean PR from a branch extracted from insightmanager_pr and not a few commits and pulls after.

File: mediation/src/main/java/net/pubnative/mediation/insights/PubnativeInsightsManager.java (hunk 1 above)
Comment (cerberillo) [inline mediation/src/main/java/net/pubnative/mediation/insights/PubnativeInsightsManager.java]: we still need to fix this

File: mediation/src/main/java/net/pubnative/mediation/insights/PubnativeInsightsManager.java [hunk 2]
```diff
@@ -74,232 +82,51 @@ public synchronized static void trackData(Context context, String baseURL, Map<S
...
-        if (context != null && model != null) {
-            List<PubnativeInsightRequestModel> pendingList = getTrackingList(context, listKey);
-            if (pendingList == null) {
-                pendingList = new ArrayList<PubnativeInsightRequestModel>();
-            }
-            pendingList.add(model);
-            setTrackingList(context, listKey, pendingList);
+            sModel = new PubnativeInsightRequestModel(uriBuilder.build().toString(), dataModel);
+            sTrackingDataString = new Gson().toJson(sModel.dataModel);
+            sUrl = sModel.url;
```
Comment (cerberillo) [inline mediation/src/main/java/net/pubnative/mediation/insights/PubnativeInsightsManager.java]: why keeping an static reference if you're keeping a reference to the model?

File: mediation/src/main/java/net/pubnative/mediation/insights/PubnativeInsightsManager.java [hunk 3]
```diff
@@ -74,232 +82,51 @@ public synchronized static void trackData(Context context, String baseURL, Map<S
...
-        Log.v(TAG, "enqueueInsightItem");
-        if (context != null && model != null) {
-            List<PubnativeInsightRequestModel> pendingList = getTrackingList(context, listKey);
-            if (pendingList == null) {
-                pendingList = new ArrayList<PubnativeInsightRequestModel>();
-            }
-            pendingList.add(model);
-            setTrackingList(context, listKey, pendingList);
+            sModel = new PubnativeInsightRequestModel(uriBuilder.build().toString(), dataModel);
+            sTrackingDataString = new Gson().toJson(sModel.dataModel);
```
Comment (cerberillo) [inline mediation/src/main/java/net/pubnative/mediation/insights/PubnativeInsightsManager.java]: Why keeping data here if you can get it when needed from `sModel`?

File: mediation/src/main/java/net/pubnative/mediation/insights/PubnativeInsightsManager.java [hunk 4]
```diff
@@ -74,232 +82,51 @@ public synchronized static void trackData(Context context, String baseURL, Map<S
...
-
-        Log.v(TAG, "enqueueInsightItem");
-        if (context != null && model != null) {
-            List<PubnativeInsightRequestModel> pendingList = getTrackingList(context, listKey);
-            if (pendingList == null) {
-                pendingList = new ArrayList<PubnativeInsightRequestModel>();
-            }
-            pendingList.add(model);
-            setTrackingList(context, listKey, pendingList);
+            sModel = new PubnativeInsightRequestModel(uriBuilder.build().toString(), dataModel);
```
Comment (cerberillo) [inline mediation/src/main/java/net/pubnative/mediation/insights/PubnativeInsightsManager.java]: The idea of the requestModel was to be used as item for a Queue, but here you're not implementing any queue, thus you should modify your architecture using your conception, not the original one that apparently fails

File: mediation/src/main/java/net/pubnative/mediation/insights/PubnativeInsightsManager.java [hunk 5]
```diff
@@ -74,232 +82,51 @@ public synchronized static void trackData(Context context, String baseURL, Map<S
...
+        public void onPubnativeHttpRequestFail(PubnativeHttpRequest request, Exception exception) {
 
-        Log.v(TAG, "getSharedPreferencesEditor");
-        SharedPreferences.Editor result = null;
-        SharedPreferences preferences = getSharedPreferences(context);
-        if (preferences != null) {
-            result = preferences.edit();
+            Log.d(TAG, "onPubnativeHttpRequestFail");
+            // Try to send data to server again
+            sendTrackingDataToServer();
```
Comment (cerberillo) [inline mediation/src/main/java/net/pubnative/mediation/insights/PubnativeInsightsManager.java]: what happens if the user closes the app? we lose all the failed insights? This has to be TOTALLY bulletproof, a generated insight MUST reach the server at some point and cannot be lost in the middle

**Comment by cerberillo** [comment]
I'm going to close this PR meanwhile because we are not going to merge it soon

**Pull request closed without merging**